### ISSUE TITLE
NickelDBus not working with 4.37.21533

### ISSUE THREAD

**Issue by dgsiegel** [shermp/NickelDBus#15]
I recently upgraded my read to the 4.37.21533 firmware and found that NickelDBus doesn't seem to expose its DBus interface anymore:

```
# qndb 
interface not valid 
```

```
# qndb -m pfmRescanBooks
interface not valid 
failed with:  "method failed with err: org.freedesktop.DBus.Error.ServiceUnknown and message: The name com.github.shermp.nickeldbus was not provided by any .service files" 
```

```
# dbus-send --system --print-reply --dest=org.freedesktop.DBus  /org/freedesktop/DBus org.freedesktop.DBus.ListNames 
method return sender=org.freedesktop.DBus -> dest=:1.28 reply_serial=2
   array [
      string "org.freedesktop.DBus"
      string ":1.28"
      string "name.marples.roy.dhcpcd"
      string ":1.24"
      string ":1.0"
   ]
```

Could it be that we need a .service file in `/usr/share/dbus-1/services/`?

**Comment by shermp** [comment]
I just tested with my dev Glo, which is running FW 4.37.21533. It seems to work fine.

It sounds like the plugin hasn't loaded, can you check for errors via `logread`? Specifically, you'll be checking to see if it has aborted loading due to loading from a failsafe.

If this is the case, you will need to delete the failsafed `libndb` file from `/usr/local/Kobo/imageformats`, and reinstall NickelDBus. (you could also just try renaming the file back to `libndb.so`)

**Comment by dgsiegel** [comment]
You were absolutely right, somehow I must have triggered failsafe:

```
Jul  3 10:15:49 nickel: (NickelDBus) (NickelHook) initializing 'NickelDBus' (version: 0.2.0)
Jul  3 10:15:49 nickel: (NickelDBus) (NickelHook) ... desc: Observe and control Nickel over D-Bus
Jul  3 10:15:49 nickel: (NickelDBus) (NickelHook) creating failsafe
Jul  3 10:15:49 nickel: (NickelDBus) (NickelHook) ... failsafe: info: allocating memory
Jul  3 10:15:49 nickel: (NickelDBus) (NickelHook) ... failsafe: info: finding filenames
Jul  3 10:15:49 nickel: (NickelDBus) (NickelHook) ... failsafe: error: lib was loaded from the failsafe for some reason (check failed: !(d && !strcmp(d, ".failsafe")))
Jul  3 10:15:49 nickel: (NickelDBus) (NickelHook) ... error creating failsafe, returning
Jul  3 10:15:49 nickel: (NickelDBus) (NickelHook) done init
```

I've renamed `libndb`, rebooted and renamed it back to `libndb.so`. It seems to be working again, thanks!

**Comment by dgsiegel** [comment]
I think I've figured out what causes this.

It seems that whenever my Kobo crashes and I have to force-reboot it, NickelDBus enters its failsafe mode. After that, I get the `interface not valid` / `org.freedesktop.DBus.Error.ServiceUnknown` errors described above because `libndb.so` has been renamed.

I can recover by renaming the library back to `libndb.so` and rebooting, so it's not a major issue.

Is there a better way to handle this? It seems like NickelDBus assumes the forced reboot was caused by itself, even if the crash was unrelated. I'm wondering if there's a way to avoid triggering the failsafe in that situation, or if this is simply the intended behavior.
